src/ml/neural_net/combine_base.hpp

```diff
@@ -0,0 +1,253 @@
+/* Copyright © 2020 Apple Inc. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-3-clause license that can
+ * be found in the LICENSE.txt file or at
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#ifndef ML_NEURAL_NET_COMBINE_BASE_HPP_
+#define ML_NEURAL_NET_COMBINE_BASE_HPP_
+
+/**
+ * \file combine_base.hpp
+ *
+ * Defines the core data types for a reactive-streams library inspired by the
+ * Swift Combine framework. Client code should generally import combine.hpp.
+ */
+
+#include <exception>
+#include <memory>
+
+namespace turi {
+namespace neural_net {
+
+// Forward declarations for types defined by other headers included by
+// combine.hpp.
+
+template <typename T, typename Callable>
+class CallableTransform;
+
+template <typename T>
+class FuturesStream;
+
+template <typename T>
+class FuturesSubscriber;
+
+template <typename T, typename U>
+class MapPublisher;
+
+template <typename T, typename U>
+class Transform;
+
+/**
+ * Simple type expressing how many values a Subscriber is ready to receive
+ * from its Publisher.
+ */
+class Demand {
+ public:
+  static Demand Unlimited() { return Demand(-1); }
+  static Demand None() { return Demand(0); }
+
+  /** Any negative value is interpreted as "unlimited". */
+  explicit Demand(int max) : max_(max) {}
+
+  bool IsUnlimited() const { return max_ < 0; }
+  bool IsNone() const { return max_ == 0; }
+
+  /** Returns a negative number to indicate "unlimited." */
+  int max() const { return max_; }
```

Comment:
Curiosity, is this method lower case since it's pretty much referencing an private member variable.

Reply:
Good question. The Google style guide makes an exception for getters/setters that are conceptually used like variables: https://google.github.io/styleguide/cppguide.html#Function_Names